Test code blocks in docs with test-doc-blocks

lread opened the pull request:
Resolves #290

**Build**

New commands:
- `gen-doc-tests` - only regenerates tests if stale,
   use `clean` command to force regen
- `run-doc-tests` - calls gen-doc-tests then runs tests,
  accepts the same parameters as run-tests.
  Can specify `:platform`
    - `:cljs` - run tests under ClojureScript
    - otherwise Clojure where we can specify one of: `:1.9`
    `:1.10` `:master`

I'm not sure if my use of the `:platform` parameter jives with
your `:aliases` parameter used for `run-tests`.
Can adjust if you like.

Example usages:
```shell
clojure -T:build gen-doc-tests
clojure -T:build run-doc-tests
clojure -T:build run-doc-tests :aliases '[:cljs]'
clojure -T:build run-doc-tests :aliases '[:1.9]'
clojure -T:build run-doc-tests :aliases '[:1.10]'
clojure -T:build run-doc-tests :aliases '[:master]'
```

The `ci` command has been updated to generate and run doc tests for same
platforms as unit tests.

**Articles**

In addition to `README.md`, now testing doc blocks in all articles
under `doc` dir excepting `doc/operator-reference.md` which does not
have any runnable code blocks.

**Skipped**

Any code block that is intentionally not runnable has been marked to be
skipped via: `<!-- :test-doc-blocks/skip -->`.

**Consistency**

I noticed that some code blocks use REPL syntax:
```Clojure
user=> (+ 1 2 3)
6
```
and others use editor syntax:
```Clojure
(+ 1 2 3)
;;=> 6
```
some places also omit the comment for editor style:
```Clojure
(+ 1 2 3)
=> 6
```
All of this is just fine with test-doc-blocks.
I left the inconsistency as is, but can make a pass for consistency upon
request.

**HoneySQL state**

I noticed a code block that set the sql dialect was affecting other
tests. I simply restored the dialect to the default at the end of the
code block.

**Un-tweaked output**

Some code blocks had string output hand-tweaked for readability.
These have been adjusted to instead use `sql/format`'s `:pretty` option.
In some cases the output is not as readable as the hand-tweaked version.
I humbly suggest that perhaps `:pretty` output could perhaps be
improved (instead of having test-doc-blocks somehow adapt).

**Corrections**

There were very few code blocks that required fixing due to incorrect
output/code.  Please review the diffs carefully to make sure all is as
expected.

**ns requires adjustments**

Any specific case of `(ns my-ns (require [my-require :as a]))` is now
the REPL friendly `(require '[my-require :as a])`

Any missing required `requires` were added.

**ClojureScript :refer**

The HoneySQL docs seem to encourage the use of referred vars for
helpers. Although this has the con of overlaps with Clojure core vars,
it is also convenient for Clojure when using `:refer :all`.

ClojureScript does not support `:refer :all` and each var must be
specified in place of `:all`.

I have adjusted examples accordingly to work with both Clojure and
ClojureScript.

lread committed:
Test code blocks in docs with test-doc-blocks

Resolves #290

**Build**

New commands:
- `gen-doc-tests` - only regenerates tests if stale,
   use `clean` command to force regen
- `run-doc-tests` - calls gen-doc-tests then runs tests,
  accepts the same parameters as run-tests.
  Can specify `:platform`
    - `:cljs` - run tests under ClojureScript
    - otherwise Clojure where we can specify one of: `:1.9`
    `:1.10` `:master`

I'm not sure if my use of the `:platform` parameter jives with
your `:aliases` parameter used for `run-tests`.
Can adjust if you like.

Example usages:
```shell
clojure -T:build gen-doc-tests

clojure -T:build run-doc-tests :platform :cljs

clojure -T:build run-doc-tests

clojure -T:build run-doc-tests :platform :1.10
```

The `ci` command has been updated to generate and run doc tests for same
platforms as unit tests.

**Articles**

In addition to `README.md`, now testing doc blocks in all articles
under `doc` dir excepting `doc/operator-reference.md` which does not
have any runnable code blocks.

**Skipped**

Any code block that is intentionally not runnable has been marked to be
skipped via: `<!-- :test-doc-blocks/skip -->`.

**Consistency**

I noticed that some code blocks use REPL syntax:
```Clojure
user=> (+ 1 2 3)
6
```
and others use editor syntax:
```Clojure
(+ 1 2 3)
;;=> 6
```
some places also omit the comment for editor style:
```Clojure
(+ 1 2 3)
=> 6
```
All of this is just fine with test-doc-blocks.
I left the inconsistency as is, but can make a pass for consistency upon
request.

**HoneySQL state**

I noticed a code block that set the sql dialect was affecting other
tests. I simply restored the dialect to the default at the end of the
code block.

**Un-tweaked output**

Some code blocks had string output hand-tweaked for readability.
These have been adjusted to instead use `sql/format`'s `:pretty` option.
In some cases the output is not as readable as the hand-tweaked version.
I humbly suggest that perhaps `:pretty` output could perhaps be
improved (instead of having test-doc-blocks somehow adapt).

**Corrections**

There were very few code blocks that required fixing due to incorrect
output/code.  Please review the diffs carefully to make sure all is as
expected.

**refer-clojure :excludes**

Not currently supported for test-doc-blocks, not a real issue for
Clojure, we'll see warnings under Clojure, but that's probably ok.

But I might actually need it for ClojureScript.
I was finding that `for` did not get overridden by our helper
`:refer` in CloureScript.

Will add proper support to test-doc-blocks but in the short-term,
will use `h/for`.

**ns requires adjustments**

Any specific case of `(ns my-ns (require [my-require :as a]))` is now
the REPL friendly `(require '[my-require :as a])`

Any missing required `requires` were added.

The HoneySQL docs seem to encourage the use of referred vars for
helpers. Although this has the con of overlaps with Clojure core vars,
it is also convenient for Clojure when using `:refer :all`.

**ClojureScript :refer**

ClojureScript does not support `:refer :all` and each var must be
specified in place of `:all`.

I have adjusted examples accordingly to work with both Clojure and
ClojureScript.

lread commented:
Hiya @seancorfield, my first crack at it.
Happy to discuss/adjust. Lemme know!

lread committed:
Merge remote-tracking branch 'upstream/develop' into lread-issue-290

Adapt to build changes.

* upstream/develop:
  Switch to build-clj lib

lread committed:
Doc block testing now understands refer-clojure

Test-doc-blocks added support for this feature

lread committed:
Tweak tests regeneration from docs

Tests will now always regenerate when:

- there is no last successful generation run
- OR when any docs, build files, or source files are newer than the
last successful generation run.

seancorfield commented:
I would prefer that the gen test doc step be run as a task (in a separate process, as Eastwood and the tests are run -- and as readme is currently run) so that the `:build` alias doesn't need additional dependencies and `build.clj` doesn't have to require those additional libraries. That would mean moving that gen test doc code into a new directory and making it a `-main` so there could be a `:gen-test-doc` alias in `deps.edn` that could run it via `-M` -- then you can do `(bb/run-task opts [:gen-test-doc])`

seancorfield commented:
Not sure why `:platform` needs to be introduced given that the tests are already run under aliases for the cljs/clj-version variants?

lread commented:
> I would prefer that the gen test doc step be run as a task (in a separate process, as Eastwood and the tests are run -- and as readme is currently run) so that the `:build` alias doesn't need additional dependencies and `build.clj` doesn't have to require those additional libraries. That would mean moving that gen test doc code into a new directory and making it a `-main` so there could be a `:gen-test-doc` alias in `deps.edn` that could run it via `-M` -- then you can do `(bb/run-task opts [:gen-test-doc])`

Yep, can do.

lread commented:
> Not sure why `:platform` needs to be introduced given that the tests are already run under aliases for the cljs/clj-version variants?

Yeah... when your previous `build.clj` exposed `run-tests` on the command line (which it doesn't anymore?) I was a bit confused on how to use it.  I thought a `:platform` abstraction might be easier to understand and ran with that. 

Would you prefer a rename to `:alias`?  (I don't know if there is a need for an `:aliases` vector here.)

lread commented:
Oh while I'm thinking of it: not sure if it is important to you but `deps.edn` `:test` brings in the cognitect test runner dep.
This dep is not technically needed for `:cljs` testing but will be present because `:test` and `:cljs` are used together.

seancorfield commented:
> Would you prefer a rename to :alias? (I don't know if there is a need for an :aliases vector here.)

The existing code uses `:aliases` because that's what is used to build the basis for process execution and `run-task` operates in terms of a vector of alias names so let's keep it consistent and use `:aliases`

> not sure if it is important to you but deps.edn :test brings in the cognitect test runner dep.
This dep is not technically needed for :cljs testing but will be present because :test and :cljs are used together.

It's not important enough to want to change things around to avoid it (but it's a fair question to ask).

lread committed:
PR review feedback: move gen-doc-tests build code

Move the code that generates doc tests out from build.clj
to its own source runnable via main.

This isolates gen-doc-tests work to its own process
during the build.

I arbitrarily:
- chose 'build' as the folder for build helper sources
- namespaced our first build helper under 'honey.gen-doc-tests'

lread committed:
PR review feedback: change run-doc-tests arg to :aliases

Example usages are now:

```
clojure -T:build run-doc-tests

clojure -T:build run-doc-tests :aliases '[:cljs]'

clojure -T:build run-doc-tests :aliases '[:1.9]'

clojure -T:build run-doc-tests :aliases '[:1.10]'

clojure -T:build run-doc-tests :aliases '[:master]'
```

lread committed:
Merge remote-tracking branch 'upstream/develop' into lread-issue-290

* upstream/develop:
  update eastwood etc
  add distribution (jdk)
  add maven cache to GH Actions

lread commented:
So @seancorfield, I think ae29147a2229193e13a4d0871c61be8e6ee6a41e is kind of what you were thinking, but details might need to be tweaked to your liking.

seancorfield commented:
That looks great @lread -- thank you so much for working on this!

lread commented:
> That looks great @lread -- thank you so much for working on this!

Thanks @seancorfield, it is always a pleasure with you!
Ping me if you find anything that needs love or could be improved in test-doc-blocks.
